File: base-pdf-reader/src/main/java/com/ezteam/ezpdflib/dialog/BaseDialog.java
package com.ezteam.ezpdflib.dialog;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.viewbinding.ViewBinding;


import java.util.HashMap;

public abstract class BaseDialog<BD extends ViewBinding, B extends BuilderDialog> extends Dialog {

    protected B builder;
    protected BD binding;

    public BaseDialog(@NonNull Context context, B builder) {
        super(context);
        this.builder = builder;
    }

    protected abstract BD getViewBinding();


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initView();
        initData();
        initListener();
    }

    protected void initView() {
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        binding = getViewBinding();
        setContentView(binding.getRoot());
        Window window = getWindow();
        WindowManager.LayoutParams windowParams = window.getAttributes();
        window.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
        windowParams.dimAmount = 0.7f;
        window.setAttributes(windowParams);
        setCancelable(true);
        setCanceledOnTouchOutside(true);
        window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
    }


    protected void initData() {
        if (!TextUtils.isEmpty(builder.title) && getTitle() != null) {
            getTitle().setVisibility(View.VISIBLE);
            getTitle().setText(builder.title);
        }

        if (!TextUtils.isEmpty(builder.message) && getMessage() != null) {
            getMessage().setVisibility(View.VISIBLE);
            getMessage().setText(builder.message);
        }

        if (!TextUtils.isEmpty(builder.positiveButton) && getPositiveButton() != null) {
            getPositiveButton().setVisibility(View.VISIBLE);
            getPositiveButton().setText(builder.positiveButton);
            getPositiveButton().setOnClickListener(v -> {
                builder.dismissDialogListener = null;
                handleClickPositiveButton(new HashMap<>());
            });
        }

        if (!TextUtils.isEmpty(builder.negativeButton) && getNegativeButton() != null) {
            getNegativeButton().setVisibility(View.VISIBLE);
            getNegativeButton().setText(builder.negativeButton);
            getNegativeButton().setOnClickListener(v -> {
                builder.dismissDialogListener = null;
                handleClickNegativeButton();
            });
        }

    }

    @Override
    public void dismiss() {
        super.dismiss();
        if (builder.dismissDialogListener != null)
            builder.dismissDialogListener.onDismissDialogListener();
    }

    protected abstract void initListener();

    protected TextView getPositiveButton() {
        return null;
    }

    protected TextView getNegativeButton() {
        return null;
    }

    protected TextView getTitle() {
        return null;
    }

    protected TextView getMessage() {
        return null;
    }

    protected void handleClickNegativeButton() {
        if (builder.negativeButtonListener != null)
            builder.negativeButtonListener.onNegativeButtonListener(this);
    }

    protected void handleClickPositiveButton(HashMap<String, Object> datas) {
        if (builder.positiveButtonListener != null) {
            builder.positiveButtonListener.onPositiveButtonListener(this
                    , datas);
        }
    }

}

File: base-module/src/main/java/com/ezteam/baseproject/dialog/BaseDialog.kt
package com.ezteam.baseproject.dialog

import android.R
import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.View
import android.view.Window
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.OnApplyWindowInsetsListener
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.fragment.app.DialogFragment
import androidx.viewbinding.ViewBinding

abstract class BaseDialog<BD : ViewBinding, B : BuilderDialog>(var builder: B, context: Context) :
    Dialog(context) {

    protected abstract val viewBinding: BD
    lateinit var binding: BD

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        binding = viewBinding
        setContentView(binding.root)
        window?.apply {
            val windowParams = attributes
            setLayout(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )
            windowParams.dimAmount = 0.7f
            attributes = windowParams
            setCancelable(true)
            setCanceledOnTouchOutside(true)
            setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        }
        initView()
        initListener()
    }

    override fun show() {
        super.show()
        setOnDismissListener {
            builder.dismissDialogListener?.invoke()
        }
        applyImmersiveMode()
    }

    private fun applyImmersiveMode() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            window?.insetsController?.let { controller ->
                controller.hide(
                    android.view.WindowInsets.Type.navigationBars())
                controller.systemBarsBehavior =
                    android.view.WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            }
        } else {
            @Suppress("DEPRECATION")
            window?.decorView?.systemUiVisibility =
                (View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                        or View.SYSTEM_UI_FLAG_FULLSCREEN
                        or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
        }
    }

    open fun initView() {
        title?.let {
            it.text = builder.title.orEmpty()
        }

        positiveButton?.let {
            if (!builder.positiveButton.isNullOrEmpty()) {
                it.text = builder.positiveButton
                it.setOnClickListener {
                    handleClickPositiveButton(HashMap())
                }
            }
        }

        negativeButton?.let {
            if (!builder.negativeButton.isNullOrEmpty()) {
                it.text = builder.negativeButton
                it.setOnClickListener(::handleClickNegativeButton)
            }
        }

        message?.let {
            it.text = builder.message.orEmpty()
        }

        container?.let {
            it.setOnClickListener {
                dismiss()
            }
        }
    }

    protected abstract fun initListener()
    protected open val positiveButton: TextView?
        get() = null
    protected open val negativeButton: TextView?
        get() = null
    protected open val title: TextView?
        get() = null
    protected open val message: TextView?
        get() = null
    protected open val container: View?
        get() = null

    protected open fun handleClickNegativeButton(view: View) {
        builder.negativeButtonListener?.let {
            it(this)
        }
        dismiss()
    }

    protected open fun handleClickPositiveButton(data: HashMap<String?, Any?>) {
        builder.positiveButtonListener?.let {
            it(this, data)
        }
        dismiss()
    }
}